apollo-federation/src/sources/connect/json_selection/selection_set.rs

```diff
@@ -0,0 +1,501 @@
+                    }
+                }
+                NamedSelection::Group(alias, sub) => {
+                    let key = alias.name.to_string();
+                    if let Some(fields) = field_map.get_vec(&key) {
+                        for field in fields {
+                            new_selections.push(NamedSelection::Group(
+                                Alias {
+                                    name: field.response_key().to_string(),
+                                },
+                                sub.apply_selection_set(&field.selection_set),
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+        let new_star = self.star.as_ref().cloned();
+        if new_star.is_some() {
+            // Alias fields that were dropped from the original selection to prevent them from
+            // being picked up by the star.
+            dropped_fields.retain(|key, _| !referenced_fields.contains(key));
+            for (dropped, _) in dropped_fields {
+                let mut unused = "__unused__".to_owned();
+                unused.push_str(dropped.as_str());
+                new_selections.push(NamedSelection::Field(
+                    Some(Alias { name: unused }),
+                    dropped.clone(),
+                    None,
+                ));
+            }
+        }
+
+        Self {
+            selections: new_selections,
+            star: new_star,
+        }
+    }
+}
+
+impl ApplySelectionSet for PathSelection {
+    fn apply_selection_set(&self, selection_set: &SelectionSet) -> Self {
+        match self {
+            Self::Var(str, path) => Self::Var(
+                str.clone(),
+                Box::new(path.apply_selection_set(selection_set)),
+            ),
+            Self::Key(key, path) => Self::Key(
+                key.clone(),
+                Box::new(path.apply_selection_set(selection_set)),
+            ),
+            Self::Selection(sub) => Self::Selection(sub.apply_selection_set(selection_set)),
+            Self::Empty => Self::Empty,
+        }
+    }
+}
+
+#[inline]
+fn key_name(path_selection: &PathSelection) -> Option<String> {
+    match path_selection {
+        PathSelection::Key(Key::Field(name), _) => Some(name.clone()),
+        PathSelection::Key(Key::Quoted(name), _) => Some(name.clone()),
+        _ => None,
+    }
+}
+
+fn map_fields_by_name(set: &SelectionSet) -> MultiMap<String, &Node<Field>> {
+    let mut map = MultiMap::new();
+    map_fields_by_name_impl(set, &mut map);
+    map
+}
+
+fn map_fields_by_name_impl<'a>(set: &'a SelectionSet, map: &mut MultiMap<String, &'a Node<Field>>) {
+    for selection in &set.selections {
+        match selection {
+            Selection::Field(field) => {
+                map.insert(field.name.to_string(), field);
+            }
+            Selection::FragmentSpread(_) => {
+                // Ignore - these are always expanded into inline fragments
```

Comment:
It's up to the caller, so we can't really guarantee it. The expansion [happens here currently](https://github.com/apollographql/router/blob/connectors/apollo-federation/src/operation/mod.rs#L2040). This could log, or panic, or return an error, instead of silently ignoring it. panic is obviously bad, an error is slightly less bad but requires all callers to handle an error that should never happen, and logging would likely go unnoticed. An error seems the least bad of the three?

Reply:
Another option would be to just handle fragment spreads and have that be dead code for now. Or to do a bigger refactor and use a type-state pattern where a `SelectionSet` is transformed into a `NormalizedSelectionSet` that doesn't have any named fragments and then take that as input here. I like that the best, but probably overkill for now.